Feature: Data Store Migration - Payments

alexflorisca opened the pull request:
This is the feature branch for the payments side of the data store migration project. This will be merged into [Feature: Data Store Migration](https://github.com/woocommerce/woocommerce-blocks/pull/6612#)

The following PRs should be reviewed and merged into this one (the order is important):

- [x] https://github.com/woocommerce/woocommerce-blocks/pull/6208
- [x] https://github.com/woocommerce/woocommerce-blocks/pull/6587
- [x] https://github.com/woocommerce/woocommerce-blocks/pull/6588
- [x] https://github.com/woocommerce/woocommerce-blocks/pull/6589
- [x] https://github.com/woocommerce/woocommerce-blocks/pull/6607

#### Other Checks

- [ ] This PR adds/removes a feature flag & I've updated [this doc](../docs/blocks/feature-flags-and-experimental-interfaces.md) .
- [ ] This PR adds/removes an experimental interfaces and I've updated [this doc](../docs/blocks/feature-flags-and-experimental-interfaces.md)
- [x] I tagged two reviewers because this PR makes queries to the database or I think it might have some security impact.

### Testing

#### Automated Tests
* [x] Changes in this PR are covered by Automated Tests.
  * [x] Unit tests
  * [x] E2E tests

#### Manual Testing

### Manual Testing

How to test the changes in this Pull Request:

#### Default payments

1. Activate `Direct bank transfer`, `Check payments`, and/or `Cash on delivery` default payment methods
2. Add an item to the cart and go to checkout
3. Choose one of the above payment methods and place an order. (the order should be successful)

#### Stripe payment

1. Install the stripe gateway plugin and activate it
2. Add an item to the cart and go to checkout
3. Enter 4242 4242 4242 4242 as a card number in the stripe payment method, valid expiry date, and any CVC code, and save the card info.
4. Place order. (the order should be successful)

#### Stripe payment (non saved card)

The `Stripe payment` is required.

1. Add an item to the cart and go to checkout
2. Enter a new Credit Card (Stripe) info (e.g., change valid expiry date or CVC code). Do not save this new card.
3. Place an order. (the order should be successful)

#### Express payment method

1. Set up an express payment method
2. Go to checkout.
3. place order.

opr committed:
Move paymentMethodDataProvider into a data store (#6208)

* Fix typographical error on LegacyRegisterExpressPaymentMethodFunction type

* Add default state for PaymentMethod data store

* Add preliminary action types

* Add preliminary action dispatchers

* Create payment method data store

* Add preliminary reducers for payment method data store

* Add preliminary selectors for payment method data store

* Add reducers/actions for registering payment methods

* Export payment method data store key

* Add test for payment method data reducers

* Add shouldSavePaymentMethod selector

* Add store key as constant

* Add more action types for registering and initializing payment methods

* Get active payment method from data store instead of from context

* Add registered methods to default state of payment method data store

* Dispatch name of registered payment method to payment method data store

* Remove setShouldSavePayment from payment method dispatcher and types

* Get payment methods from registry instead of payment context

* Add available payment methods to store

* Add function to check whether payment methods are allowed to be used

* Add selector to check if payments are initialised

* Remove resolvers and add controls to payment method data store

* Change type of payment requirements to string[]

* Turn addRegistered and addRegisteredExpress into generators

This is so we can check each payment method's validity before adding it to the list of available payment methods

* Add action type for setting express payments as initialized

* Only select from available methods in payment method options

* Remove argument from addRegisteredPaymentMethod in payment method registry

* Rename folder and store name to not contain the word data

* Add selectors for express payment methods and their initialisation

* Delete controls again in favour of thunks

* Rename payment-method-data to payment-methdods

* Create new setDefaultPaymentMethod function

This will set the payment method when the cart loads.

* Add CustomerPaymentMethodConfiguration type

* Make getAvailableExpressPaymentMethods return correct data

* Check express methods and normal methods when cart changes

* Add action for setting active express payment methods

* Handle express methods in checkPaymentMethodCanPay

* Hide express payments area if none are available

* Add selector for paymentMethodData

* Add customer data to default state and add selector for it

* Add setPaymentStatus action and reducer case

* Set the default payment method when one isn't selected

* Correct types on getCustomerPaymentMethods

* Set status in data store alongside context status

* Comment out active gateway selection - remove later

* Set status in express payment methods in data store

* Use correct key in payment method data context

* Directly check payment methods from the list in blocks-registry

* Add semicolon to import statement

* Fix payment method data state call

* Get paymentMethodData from store not context

* Fix linting issues

Co-authored-by: Saad Tarhi <saad.trh@gmail.com>
Co-authored-by: Alex Florisca <alex.florisca@automattic.com>

tarhi-saad committed:
Rebase on the update/checkout-data-store branch & Fix failed payments (#6587)

* Add checkout data store

* wip on checkout data store

* CheckoutContext now uses the checkout store

* Investigated and removed setting the redirectUrl on the default state

* update extension and address hooks to use checkout data store

* use checkout data store in checkout-processor and use-checkout-button

* trim useCheckoutContext from use-payment-method-interface && use-store-cart-item-quantity

* Remove useCheckoutContext from shipping provider

* Remove isCalculating from state

* Removed useCheckoutContext from lots of places

* Remove useCheckoutContext from checkout-payment-block

* Remove useCheckoutContext in checkout-shipping-methods-block and checkout-shipping-address-block

* add isCart selector and action and update the checkoutstate context

* Fixed redirectUrl bug by using thunks

* Remove dispatchActions from checkout-state

* Change SET_HAS_ERROR action to be neater

* Thomas' feedback

* Tidy up

* Oops, deleted things I shouldn't have

* Typescript

* Fix types

* Fix tests

* Remove isCart

* Update docs and remove unecessary getRedirectUrl() selector

* set correct type for preloadedCheckoutData

* Remove duplicate Address type

* Fix missing addresses from type-defs index

* Update docs/block-client-apis/checkout/checkout-api.md

Co-authored-by: Thomas Roberts <5656702+opr@users.noreply.github.com>

* Update docs/block-client-apis/checkout/checkout-api.md

Co-authored-by: Thomas Roberts <5656702+opr@users.noreply.github.com>

* Update docs

* Update docs/block-client-apis/checkout/checkout-api.md

Co-authored-by: Thomas Roberts <5656702+opr@users.noreply.github.com>

* Update docs/block-client-apis/checkout/checkout-api.md

Co-authored-by: Thomas Roberts <5656702+opr@users.noreply.github.com>

* Revert feedback changes

* REvert feedback formatting

* Update docs formatting

* Fix typographical error on LegacyRegisterExpressPaymentMethodFunction type

* Add default state for PaymentMethod data store

* Add preliminary action types

* Add preliminary action dispatchers

* Create payment method data store

* Add preliminary reducers for payment method data store

* Add preliminary selectors for payment method data store

* Add reducers/actions for registering payment methods

* Export payment method data store key

* Add test for payment method data reducers

* Add shouldSavePaymentMethod selector

* Add store key as constant

* Add more action types for registering and initializing payment methods

* Get active payment method from data store instead of from context

* Add registered methods to default state of payment method data store

* Dispatch name of registered payment method to payment method data store

* Remove setShouldSavePayment from payment method dispatcher and types

* Get payment methods from registry instead of payment context

* Add available payment methods to store

* Add function to check whether payment methods are allowed to be used

* Add selector to check if payments are initialised

* Remove resolvers and add controls to payment method data store

* Change type of payment requirements to string[]

* Turn addRegistered and addRegisteredExpress into generators

This is so we can check each payment method's validity before adding it to the list of available payment methods

* Add action type for setting express payments as initialized

* Only select from available methods in payment method options

* Remove argument from addRegisteredPaymentMethod in payment method registry

* Rename folder and store name to not contain the word data

* Add selectors for express payment methods and their initialisation

* Delete controls again in favour of thunks

* Rename payment-method-data to payment-methdods

* Create new setDefaultPaymentMethod function

This will set the payment method when the cart loads.

* Add CustomerPaymentMethodConfiguration type

* Make getAvailableExpressPaymentMethods return correct data

* Check express methods and normal methods when cart changes

* Add action for setting active express payment methods

* Handle express methods in checkPaymentMethodCanPay

* Hide express payments area if none are available

* Add selector for paymentMethodData

* Add customer data to default state and add selector for it

* Add setPaymentStatus action and reducer case

* Set the default payment method when one isn't selected

* Correct types on getCustomerPaymentMethods

* Set status in data store alongside context status

* Comment out active gateway selection - remove later

* Set status in express payment methods in data store

* Directly check payment methods from the list in blocks-registry

* Add semicolon to import statement

* Fix payment method data state call

* Get paymentMethodData from store not context

* Add addPaymentMethodData action/reducer case

* Update payment method on payment success

* Add 'getCurrentStatus' selector

* Remove the temporary solution

For getting payment method data into the data store

* Prevent the 'success' context action from being dispatched

* Update the "setPaymentStatus" data store action

Accept status as an object instead of string

* Fix the "currentStatus" reducer state update value

* Get payment data into data store

* Set the correct payment status to data store

* Get the success status of payment from data store

* Use store data in the payment dispatchers

Replace the React useReducers action in the payment dispatchers file with
the payment method data store

* Get payment status from data store

* Use data store for the payment error status

* Use payment data store failed status

* Use payment data store for the isFinished status

Co-authored-by: Alex Florisca <alex.florisca@automattic.com>
Co-authored-by: Thomas Roberts <5656702+opr@users.noreply.github.com>
Co-authored-by: Thomas Roberts <thomas.roberts@automattic.com>

Commit:
Fix merge conflict error

tarhi-saad committed:
Set & get the provider's state from our data store instead of React's useReducer (#6588)

* Add checkout data store

* wip on checkout data store

* CheckoutContext now uses the checkout store

* Investigated and removed setting the redirectUrl on the default state

* update extension and address hooks to use checkout data store

* use checkout data store in checkout-processor and use-checkout-button

* trim useCheckoutContext from use-payment-method-interface && use-store-cart-item-quantity

* Remove useCheckoutContext from shipping provider

* Remove isCalculating from state

* Removed useCheckoutContext from lots of places

* Remove useCheckoutContext from checkout-payment-block

* Remove useCheckoutContext in checkout-shipping-methods-block and checkout-shipping-address-block

* add isCart selector and action and update the checkoutstate context

* Fixed redirectUrl bug by using thunks

* Remove dispatchActions from checkout-state

* Change SET_HAS_ERROR action to be neater

* Thomas' feedback

* Tidy up

* Oops, deleted things I shouldn't have

* Typescript

* Fix types

* Fix tests

* Remove isCart

* Update docs and remove unecessary getRedirectUrl() selector

* set correct type for preloadedCheckoutData

* Remove duplicate Address type

* Fix missing addresses from type-defs index

* Update docs/block-client-apis/checkout/checkout-api.md

Co-authored-by: Thomas Roberts <5656702+opr@users.noreply.github.com>

* Update docs/block-client-apis/checkout/checkout-api.md

Co-authored-by: Thomas Roberts <5656702+opr@users.noreply.github.com>

* Update docs

* Update docs/block-client-apis/checkout/checkout-api.md

Co-authored-by: Thomas Roberts <5656702+opr@users.noreply.github.com>

* Update docs/block-client-apis/checkout/checkout-api.md

Co-authored-by: Thomas Roberts <5656702+opr@users.noreply.github.com>

* Revert feedback changes

* REvert feedback formatting

* Update docs formatting

* Fix typographical error on LegacyRegisterExpressPaymentMethodFunction type

* Add default state for PaymentMethod data store

* Add preliminary action types

* Add preliminary action dispatchers

* Create payment method data store

* Add preliminary reducers for payment method data store

* Add preliminary selectors for payment method data store

* Add reducers/actions for registering payment methods

* Export payment method data store key

* Add test for payment method data reducers

* Add shouldSavePaymentMethod selector

* Add store key as constant

* Add more action types for registering and initializing payment methods

* Get active payment method from data store instead of from context

* Add registered methods to default state of payment method data store

* Dispatch name of registered payment method to payment method data store

* Remove setShouldSavePayment from payment method dispatcher and types

* Get payment methods from registry instead of payment context

* Add available payment methods to store

* Add function to check whether payment methods are allowed to be used

* Add selector to check if payments are initialised

* Remove resolvers and add controls to payment method data store

* Change type of payment requirements to string[]

* Turn addRegistered and addRegisteredExpress into generators

This is so we can check each payment method's validity before adding it to the list of available payment methods

* Add action type for setting express payments as initialized

* Only select from available methods in payment method options

* Remove argument from addRegisteredPaymentMethod in payment method registry

* Rename folder and store name to not contain the word data

* Add selectors for express payment methods and their initialisation

* Delete controls again in favour of thunks

* Rename payment-method-data to payment-methdods

* Create new setDefaultPaymentMethod function

This will set the payment method when the cart loads.

* Add CustomerPaymentMethodConfiguration type

* Make getAvailableExpressPaymentMethods return correct data

* Check express methods and normal methods when cart changes

* Add action for setting active express payment methods

* Handle express methods in checkPaymentMethodCanPay

* Hide express payments area if none are available

* Add selector for paymentMethodData

* Add customer data to default state and add selector for it

* Add setPaymentStatus action and reducer case

* Set the default payment method when one isn't selected

* Correct types on getCustomerPaymentMethods

* Set status in data store alongside context status

* Comment out active gateway selection - remove later

* Set status in express payment methods in data store

* Directly check payment methods from the list in blocks-registry

* Add semicolon to import statement

* Fix payment method data state call

* Get paymentMethodData from store not context

* Add addPaymentMethodData action/reducer case

* Update payment method on payment success

* Add 'getCurrentStatus' selector

* Remove the temporary solution

For getting payment method data into the data store

* Prevent the 'success' context action from being dispatched

* Update the "setPaymentStatus" data store action

Accept status as an object instead of string

* Fix the "currentStatus" reducer state update value

* Get payment data into data store

* Set the correct payment status to data store

* Get the success status of payment from data store

* Use store data in the payment dispatchers

Replace the React useReducers action in the payment dispatchers file with
the payment method data store

* Get payment status from data store

* Use data store for the payment error status

* Use payment data store failed status

* Use payment data store for the isFinished status

* Update the setPaymentStatus argument

* Set up setRegisteredPaymentMethods in data store

* Use the data store version of the registeredPaymentMethods

* Fix the default state type of the registeredPaymentMethods

* Set up setRegisteredExpressPaymentMethod in data store

* Use the data version of the registred express payment methods

* Set the correct action type for removing payment methods

* Fix default state express payment methods type

* Use the store data version of activePaymentMethod

* Use setActivePaymentMethod in the dispatchers file

And refactor code

* Update payment status arguments for express payment methods

* Use paymentMethodData from the data store

* Use payment method's errorMessage from data store

* Update paymentMethods list in data store reducer

* Replace remaining payment context data with data store

* Clean up payment method context file

Co-authored-by: Alex Florisca <alex.florisca@automattic.com>
Co-authored-by: Thomas Roberts <5656702+opr@users.noreply.github.com>
Co-authored-by: Thomas Roberts <thomas.roberts@automattic.com>

tarhi-saad committed:
Get payment method data directly from the data store instead of the usePaymentMethodDataContext hook (#6589)

* Add checkout data store

* wip on checkout data store

* CheckoutContext now uses the checkout store

* Investigated and removed setting the redirectUrl on the default state

* update extension and address hooks to use checkout data store

* use checkout data store in checkout-processor and use-checkout-button

* trim useCheckoutContext from use-payment-method-interface && use-store-cart-item-quantity

* Remove useCheckoutContext from shipping provider

* Remove isCalculating from state

* Removed useCheckoutContext from lots of places

* Remove useCheckoutContext from checkout-payment-block

* Remove useCheckoutContext in checkout-shipping-methods-block and checkout-shipping-address-block

* add isCart selector and action and update the checkoutstate context

* Fixed redirectUrl bug by using thunks

* Remove dispatchActions from checkout-state

* Change SET_HAS_ERROR action to be neater

* Thomas' feedback

* Tidy up

* Oops, deleted things I shouldn't have

* Typescript

* Fix types

* Fix tests

* Remove isCart

* Update docs and remove unecessary getRedirectUrl() selector

* set correct type for preloadedCheckoutData

* Remove duplicate Address type

* Fix missing addresses from type-defs index

* Update docs/block-client-apis/checkout/checkout-api.md

Co-authored-by: Thomas Roberts <5656702+opr@users.noreply.github.com>

* Update docs/block-client-apis/checkout/checkout-api.md

Co-authored-by: Thomas Roberts <5656702+opr@users.noreply.github.com>

* Update docs

* Update docs/block-client-apis/checkout/checkout-api.md

Co-authored-by: Thomas Roberts <5656702+opr@users.noreply.github.com>

* Update docs/block-client-apis/checkout/checkout-api.md

Co-authored-by: Thomas Roberts <5656702+opr@users.noreply.github.com>

* Revert feedback changes

* REvert feedback formatting

* Update docs formatting

* Fix typographical error on LegacyRegisterExpressPaymentMethodFunction type

* Add default state for PaymentMethod data store

* Add preliminary action types

* Add preliminary action dispatchers

* Create payment method data store

* Add preliminary reducers for payment method data store

* Add preliminary selectors for payment method data store

* Add reducers/actions for registering payment methods

* Export payment method data store key

* Add test for payment method data reducers

* Add shouldSavePaymentMethod selector

* Add store key as constant

* Add more action types for registering and initializing payment methods

* Get active payment method from data store instead of from context

* Add registered methods to default state of payment method data store

* Dispatch name of registered payment method to payment method data store

* Remove setShouldSavePayment from payment method dispatcher and types

* Get payment methods from registry instead of payment context

* Add available payment methods to store

* Add function to check whether payment methods are allowed to be used

* Add selector to check if payments are initialised

* Remove resolvers and add controls to payment method data store

* Change type of payment requirements to string[]

* Turn addRegistered and addRegisteredExpress into generators

This is so we can check each payment method's validity before adding it to the list of available payment methods

* Add action type for setting express payments as initialized

* Only select from available methods in payment method options

* Remove argument from addRegisteredPaymentMethod in payment method registry

* Rename folder and store name to not contain the word data

* Add selectors for express payment methods and their initialisation

* Delete controls again in favour of thunks

* Rename payment-method-data to payment-methdods

* Create new setDefaultPaymentMethod function

This will set the payment method when the cart loads.

* Add CustomerPaymentMethodConfiguration type

* Make getAvailableExpressPaymentMethods return correct data

* Check express methods and normal methods when cart changes

* Add action for setting active express payment methods

* Handle express methods in checkPaymentMethodCanPay

* Hide express payments area if none are available

* Add selector for paymentMethodData

* Add customer data to default state and add selector for it

* Add setPaymentStatus action and reducer case

* Set the default payment method when one isn't selected

* Correct types on getCustomerPaymentMethods

* Set status in data store alongside context status

* Comment out active gateway selection - remove later

* Set status in express payment methods in data store

* Directly check payment methods from the list in blocks-registry

* Add semicolon to import statement

* Fix payment method data state call

* Get paymentMethodData from store not context

* Add addPaymentMethodData action/reducer case

* Update payment method on payment success

* Add 'getCurrentStatus' selector

* Remove the temporary solution

For getting payment method data into the data store

* Prevent the 'success' context action from being dispatched

* Update the "setPaymentStatus" data store action

Accept status as an object instead of string

* Fix the "currentStatus" reducer state update value

* Get payment data into data store

* Set the correct payment status to data store

* Get the success status of payment from data store

* Use store data in the payment dispatchers

Replace the React useReducers action in the payment dispatchers file with
the payment method data store

* Get payment status from data store

* Use data store for the payment error status

* Use payment data store failed status

* Use payment data store for the isFinished status

* Update the setPaymentStatus argument

* Set up setRegisteredPaymentMethods in data store

* Use the data store version of the registeredPaymentMethods

* Fix the default state type of the registeredPaymentMethods

* Set up setRegisteredExpressPaymentMethod in data store

* Use the data version of the registred express payment methods

* Set the correct action type for removing payment methods

* Fix default state express payment methods type

* Use the store data version of activePaymentMethod

* Use setActivePaymentMethod in the dispatchers file

And refactor code

* Update payment status arguments for express payment methods

* Use paymentMethodData from the data store

* Use payment method's errorMessage from data store

* Update paymentMethods list in data store reducer

* Replace remaining payment context data with data store

* Clean up payment method context file

* Get payment method state from data store in the checkout submit hook

* Copy types.ts file into the payment data store folder

* Fix isExpressPaymentMethodActive selector

* Move the entire currentStatus into the data store

* Replace the payment context state with the data store

* Fix getActiveSavedToken & clean up the context file

* Use the accutrate name of the "createErrorNotice"

* Update the payment method data store key import

* Diable unused state from the context

Co-authored-by: Alex Florisca <alex.florisca@automattic.com>
Co-authored-by: Thomas Roberts <5656702+opr@users.noreply.github.com>
Co-authored-by: Thomas Roberts <thomas.roberts@automattic.com>

tarhi-saad committed:
Refactor the payment method data store & context (#6607)

* Add checkout data store

* wip on checkout data store

* CheckoutContext now uses the checkout store

* Investigated and removed setting the redirectUrl on the default state

* update extension and address hooks to use checkout data store

* use checkout data store in checkout-processor and use-checkout-button

* trim useCheckoutContext from use-payment-method-interface && use-store-cart-item-quantity

* Remove useCheckoutContext from shipping provider

* Remove isCalculating from state

* Removed useCheckoutContext from lots of places

* Remove useCheckoutContext from checkout-payment-block

* Remove useCheckoutContext in checkout-shipping-methods-block and checkout-shipping-address-block

* add isCart selector and action and update the checkoutstate context

* Fixed redirectUrl bug by using thunks

* Remove dispatchActions from checkout-state

* Change SET_HAS_ERROR action to be neater

* Thomas' feedback

* Tidy up

* Oops, deleted things I shouldn't have

* Typescript

* Fix types

* Fix tests

* Remove isCart

* Update docs and remove unecessary getRedirectUrl() selector

* set correct type for preloadedCheckoutData

* Remove duplicate Address type

* Fix missing addresses from type-defs index

* Update docs/block-client-apis/checkout/checkout-api.md

Co-authored-by: Thomas Roberts <5656702+opr@users.noreply.github.com>

* Update docs/block-client-apis/checkout/checkout-api.md

Co-authored-by: Thomas Roberts <5656702+opr@users.noreply.github.com>

* Update docs

* Update docs/block-client-apis/checkout/checkout-api.md

Co-authored-by: Thomas Roberts <5656702+opr@users.noreply.github.com>

* Update docs/block-client-apis/checkout/checkout-api.md

Co-authored-by: Thomas Roberts <5656702+opr@users.noreply.github.com>

* Revert feedback changes

* REvert feedback formatting

* Update docs formatting

* Fix typographical error on LegacyRegisterExpressPaymentMethodFunction type

* Add default state for PaymentMethod data store

* Add preliminary action types

* Add preliminary action dispatchers

* Create payment method data store

* Add preliminary reducers for payment method data store

* Add preliminary selectors for payment method data store

* Add reducers/actions for registering payment methods

* Export payment method data store key

* Add test for payment method data reducers

* Add shouldSavePaymentMethod selector

* Add store key as constant

* Add more action types for registering and initializing payment methods

* Get active payment method from data store instead of from context

* Add registered methods to default state of payment method data store

* Dispatch name of registered payment method to payment method data store

* Remove setShouldSavePayment from payment method dispatcher and types

* Get payment methods from registry instead of payment context

* Add available payment methods to store

* Add function to check whether payment methods are allowed to be used

* Add selector to check if payments are initialised

* Remove resolvers and add controls to payment method data store

* Change type of payment requirements to string[]

* Turn addRegistered and addRegisteredExpress into generators

This is so we can check each payment method's validity before adding it to the list of available payment methods

* Add action type for setting express payments as initialized

* Only select from available methods in payment method options

* Remove argument from addRegisteredPaymentMethod in payment method registry

* Rename folder and store name to not contain the word data

* Add selectors for express payment methods and their initialisation

* Delete controls again in favour of thunks

* Rename payment-method-data to payment-methdods

* Create new setDefaultPaymentMethod function

This will set the payment method when the cart loads.

* Add CustomerPaymentMethodConfiguration type

* Make getAvailableExpressPaymentMethods return correct data

* Check express methods and normal methods when cart changes

* Add action for setting active express payment methods

* Handle express methods in checkPaymentMethodCanPay

* Hide express payments area if none are available

* Add selector for paymentMethodData

* Add customer data to default state and add selector for it

* Add setPaymentStatus action and reducer case

* Set the default payment method when one isn't selected

* Correct types on getCustomerPaymentMethods

* Set status in data store alongside context status

* Comment out active gateway selection - remove later

* Set status in express payment methods in data store

* Directly check payment methods from the list in blocks-registry

* Add semicolon to import statement

* Fix payment method data state call

* Get paymentMethodData from store not context

* Add addPaymentMethodData action/reducer case

* Update payment method on payment success

* Add 'getCurrentStatus' selector

* Remove the temporary solution

For getting payment method data into the data store

* Prevent the 'success' context action from being dispatched

* Update the "setPaymentStatus" data store action

Accept status as an object instead of string

* Fix the "currentStatus" reducer state update value

* Get payment data into data store

* Set the correct payment status to data store

* Get the success status of payment from data store

* Use store data in the payment dispatchers

Replace the React useReducers action in the payment dispatchers file with
the payment method data store

* Get payment status from data store

* Use data store for the payment error status

* Use payment data store failed status

* Use payment data store for the isFinished status

* Update the setPaymentStatus argument

* Set up setRegisteredPaymentMethods in data store

* Use the data store version of the registeredPaymentMethods

* Fix the default state type of the registeredPaymentMethods

* Set up setRegisteredExpressPaymentMethod in data store

* Use the data version of the registred express payment methods

* Set the correct action type for removing payment methods

* Fix default state express payment methods type

* Use the store data version of activePaymentMethod

* Use setActivePaymentMethod in the dispatchers file

And refactor code

* Update payment status arguments for express payment methods

* Use paymentMethodData from the data store

* Use payment method's errorMessage from data store

* Update paymentMethods list in data store reducer

* Replace remaining payment context data with data store

* Clean up payment method context file

* Get payment method state from data store in the checkout submit hook

* Copy types.ts file into the payment data store folder

* Fix isExpressPaymentMethodActive selector

* Move the entire currentStatus into the data store

* Replace the payment context state with the data store

* Fix getActiveSavedToken & clean up the context file

* Use the accutrate name of the "createErrorNotice"

* Update the payment method data store key import

* Diable unused state from the context

* Get enabled customer payment methods using data store selector

* Remove remaining useReducer action from the dispatchers file

* Update types and remove unused vars

* Remove the payment method dispatchers hook

* Refactor & clean up (remove unused files)

* Remove commented line from payment methods types

* Move event emitter into thunks

Co-authored-by: Alex Florisca <alex.florisca@automattic.com>
Co-authored-by: Thomas Roberts <5656702+opr@users.noreply.github.com>
Co-authored-by: Thomas Roberts <thomas.roberts@automattic.com>

Commit:
Remove checkout-state after merge conflicts

Commit:
Fix linting errors

opr commented on assets/js/data/payment-methods/check-payment-methods.ts:
```diff
@@ -0,0 +1,119 @@
+			}
+		} catch ( e ) {
+			if ( CURRENT_USER_IS_ADMIN ) {
+				// TODO: When https://github.com/woocommerce/woocommerce-gutenberg-products-block/pull/6159 is merged we can interact with the data store.
```
Add the error notice via `core/notices` data store here ⭐

opr commented on docs/internal-developers/block-client-apis/checkout/checkout-api.md:
```diff
@@ -2,6 +2,14 @@
 **Note on migration:** We are in the process of moving much of the data from contexts into data stores, so this portion of the docs may change often as we do this. We will endavour to keep it up to date while the work is carried out
 
+**Note on migration:** We are in the process of moving much of the data from contexts into data stores, so this portion of the docs may change often as we do this. We will endavour to keep it up to date while the work is carried out
+
+**Note on migration:** We are in the process of moving much of the data from contexts into data stores, so this portion of the docs may change often as we do this. We will endavour to keep it up to date while the work is carried out
+
+**Note on migration:** We are in the process of moving much of the data from contexts into data stores, so this portion of the docs may change often as we do this. We will endavour to keep it up to date while the work is carried out
+
+**Note on migration:** We are in the process of moving much of the data from contexts into data stores, so this portion of the docs may change often as we do this. We will endavour to keep it up to date while the work is carried out
+
```
Duplicated data

opr commented on assets/js/data/payment-methods/utils.ts:
```diff
@@ -0,0 +1,50 @@
+/**
+ * Gets the payment methods saved for the current user after filtering out disabled ones.
+ */
+export const filterEnabledCustomerPaymentMethods = (
```
Adding an issue to create tests for this: https://github.com/woocommerce/woocommerce-blocks/issues/6629

opr commented on assets/js/base/context/hooks/payment-methods/use-payment-method-interface.ts:
```diff
@@ -49,14 +53,20 @@ export const usePaymentMethodInterface = (): PaymentMethodInterface => {
+		}
+	);
+
+	const { onPaymentProcessing, setExpressPaymentError } =
```
Can we at least move `setExpressPaymentError` into the data store in this PR?

opr commented on assets/js/data/payment-methods/actions.ts:
```diff
@@ -0,0 +1,145 @@
+	name,
+} );
+
+export function addRegisteredExpressPaymentMethod() {
```
I know I wrote this, but do you think we could rename this to better reflect how it is used? Maybe `processAndUpdateRegisteredExpressPaymentMethods` or something along those lines?

opr commented on assets/js/data/payment-methods/actions.ts:
```diff
@@ -0,0 +1,145 @@
+	methods,
+} );
+
+export function addRegisteredPaymentMethod() {
```
I know I wrote this, but do you think we could rename this to better reflect how it is used? Maybe `processAndUpdateRegisteredPaymentMethods` or something along those lines?

opr commented on assets/js/data/payment-methods/index.ts:
```diff
@@ -0,0 +1,43 @@
+	reducer,
+	selectors,
+	actions,
+	// eslint-disable-next-line @typescript-eslint/no-explicit-any
+	controls: { ...dataControls, ...sharedControls } as any,
```
How come we need to include these here?

opr commented on assets/js/data/cart/index.ts:
```diff
@@ -27,6 +28,10 @@ const registeredStore = registerStore< State >( STORE_KEY, {
 } );
 
 registeredStore.subscribe( pushChanges );
+registeredStore.subscribe( async () => {
```
Added https://github.com/woocommerce/woocommerce-blocks/issues/6630 to see if this can be improved

opr commented on assets/js/base/context/providers/cart-checkout/payment-methods/constants.ts:
```diff
@@ -23,56 +20,8 @@ export enum ACTION {
-	activePaymentMethod: '',
-	activeSavedToken: '',
-	setActivePaymentMethod: () => void null,
 	customerPaymentMethods: {},
```
Can this be removed from context?

opr commented on assets/js/blocks/cart-checkout-shared/payment-methods/express-payment-methods.js:
```diff
@@ -17,23 +17,33 @@ import {
+	const { setActivePaymentMethod, setPaymentStatus } = useDispatch(
+		PAYMENT_METHOD_DATA_STORE_KEY
+	);
+	const paymentMethods = useExpressPaymentMethods();
```
Could the data returned by this hook be derived from the data store?

opr commented on assets/js/data/payment-methods/thunks.ts:
```diff
@@ -0,0 +1,118 @@
+			registry.dispatch( 'core/notices' );
+		removeNotice( 'wc-payment-error', noticeContexts.PAYMENTS );
+		emitEventWithAbort(
+			// TODO Move all this block into a function in thunks & Create buttons to trigger this event (see Alex's PR)
```
Will [my testing notes](https://gist.github.com/opr/72f3513feb9e39d95c3db57fada31122) cover this?

opr reviewed (commented):
This is looking really nice, I've left some comments about things I noticed, I see we have a bunch of todos, which is fine. Apologies if any of my comments double-up on something we already know about.

Commit:
Move types to types.ts

Commit:
Move default states into respective store folders

alexflorisca commented on assets/js/data/payment-methods/default-state.ts:
```diff
@@ -0,0 +1,72 @@
+		| EmptyObjectType;
+}
+export const defaultPaymentMethodDataState: PaymentMethodDataState = {
+	paymentStatuses: PAYMENT_METHOD_STATUS,
```
Shouldn't this be `paymentStatus` singular?

Commit:
Fix types and add comment

Commit:
Move setExpressPaymentError to payment-methods store

Commit:
fix express payment methods not showing up

opr committed:
Check if payment method is active from the state

Commit:
Add comments

Commit:
Remove commented out code in payment method data context

opr commented on assets/js/data/payment-methods/reducers.ts:
```diff
@@ -0,0 +1,165 @@
+				...state.registeredPaymentMethods,
+			};
+			delete previousExpressRegisteredPaymentMethods[
+				action.registeredExpressPaymentMethods.name
```
I will change this, it should be `action.name`

Commit:
Display an error in the check-payment-methods directly from data store

Commit:
Remove use-emit-response hook and move utils in event-emit/utils.ts

opr committed:
Use correct action property to remove payment methods

opr committed:
Fix formatting

opr committed:
Only try to initialize payment methods when cart is done loading

opr committed:
Add function to order payment methods from server

opr committed:
Add payment methods in the correct order

opr committed:
Prevent adding registered payment methods before cart is ready

opr committed:
Ensure payment methods get removed from state when deregistered

opr committed:
Reorder setting default payment methods to add customer methods first

opr committed:
Get customer methods from store not context

Commit:
Remove error from payment-method state and associated selectors

Commit:
Merge branch 'feature/data-store-refactor-payments' of github.com:woocommerce/woocommerce-blocks into feature/data-store-refactor-payments

Commit:
Remove use-payment-method-registration and update the payment method state to remove the duplicated registeredPaymentMethods

Commit:
Remove errorMessage from payment-methods store

Commit:
Rename customerPaymentMethods -> savedPaymentMethods

Commit:
Order payment methods when validating

Commit:
Refactor payment-methods.js

tarhi-saad commented on assets/js/base/context/providers/cart-checkout/payment-methods/payment-method-data-context.tsx:
```diff
@@ -249,125 +128,44 @@ export const PaymentMethodDataProvider = ( {
-		isSuccessResponse,
-		isFailResponse,
-		isErrorResponse,
-		addErrorNotice,
```
The `emitPaymentProcessingEvent` depends on the above-removed events (see the assets/js/data/payment-methods/thunks.ts file). Removing them from the dependency array will prevent this effect re-run when it's required.

tarhi-saad commented on assets/js/blocks-registry/payment-methods/registry.ts:
```diff
@@ -123,12 +125,20 @@ export const __experimentalDeRegisterPaymentMethod = (
+	const { removeRegisteredPaymentMethod } = dispatch(
+		PAYMENT_METHOD_DATA_STORE_KEY
+	);
+	removeRegisteredPaymentMethod( paymentMethodName );
```
The `__experimentalDeRegisterPaymentMethod` isn't used anywhere. So, there is no value in updating its implementation!

tarhi-saad commented on assets/js/blocks-registry/payment-methods/registry.ts:
```diff
@@ -123,12 +125,20 @@ export const __experimentalDeRegisterPaymentMethod = (
+	const { removeRegisteredExpressPaymentMethod } = dispatch(
+		PAYMENT_METHOD_DATA_STORE_KEY
+	);
+	removeRegisteredExpressPaymentMethod( paymentMethodName );
```
The `__experimentalDeRegisterExpressPaymentMethod` isn't used anywhere. So, there is no value in updating its implementation!

tarhi-saad committed:
Fix "Payment methods not set in editor" #6655 bug

We never get to load the payment methods object in the editor mode
because there are no cart totals to load.

tarhi-saad commented on assets/js/data/payment-methods/thunks.ts:
```diff
@@ -0,0 +1,135 @@
+				registry.dispatch( CART_STORE_KEY );
+
+			if ( successResponse && ! errorResponse ) {
+				const { paymentMethodData, billingData, shippingData } =
```
We should use `billingAddress` instead of `billingData`

tarhi-saad commented on assets/js/data/payment-methods/thunks.ts:
```diff
@@ -0,0 +1,135 @@
+					} );
+				}
+
+				const { paymentMethodData, billingData } =
```
We should use `billingAddress` instead of `billingData`

tarhi-saad commented on assets/js/base/context/hooks/payment-methods/use-payment-methods.ts:
```diff
@@ -24,12 +25,19 @@ interface ExpressPaymentMethodState {
+			};
+		} );
+
+	const paymentMethods = getPaymentMethods();
```
We shouldn't use `paymentMethods` from the `getPaymentMethods` function directly. It isn't reordered and isn't filtered by keeping only the available payment methods (i.e., using `paymentMethod.canMakePayment()`).

tarhi-saad committed:
Initialize payment methods when available payments are loaded

tarhi-saad committed:
Remove duplicate code

tarhi-saad committed:
Fix data store state mutation anti-pattern

A Redux rule is to never mutate the state in a reducer to avoid any unexpected results

tarhi-saad committed:
Set availablePaymentMethods to the paymentMethods object

Instead of its keys. We can get the keys using "Object.keys".

tarhi-saad committed:
Use the available ordered payment methods

The `getPaymentMethods` & `getExpressPaymentMethods` may include unordored
& unavailable payment methods.

tarhi-saad committed:
Get the correct value from the emit event response

tarhi-saad committed:
Merge branch 'feature/data-store-refactor' into feature/data-store-refactor-payments

tarhi-saad committed:
Fix merge conflicts & clean up
